### PR TITLE
Throwing an exception when the afterSave aborts the transaction

### LINKED ISSUE #9079: Table::save() should return false if transaction gets rolled back in afterSave() callback
Currently if for e.g. you are saving a record in `afterSave()` callback and the operation fails the transaction would be rolled back but the primary `save()` would still return an entity instead of `false`.

@lorenzo is aware of the problem so I won't go into details :)

### PR DESCRIPTION
**Pull request by lorenzo**
Fixes #9079

### CODE REVIEW

**Comment by markstory** [comment]
Seems reasonable to me. Will this new code result in an exception being raised if an afterSave() listener commits the parent transaction as well?

**Comment by lorenzo** [comment]
@markstory Yes, but that is more difficult to do by accident, since you will need to commit as many times as `begin` was called... that cannot just happen by accident

**Comment by markstory** [comment]
@lorenzo Fair point.

**Comment by Xymanek** [comment]
Suggestion: add a parent database exception which controller or anything else that calls the save() can try/catch and display a nice "Database error" message

P.s. On mobile now, will provide an example if I'm not clear when I get home

**Comment by markstory** [comment]
@lorenzo Fair point.

**Comment by lorenzo** [comment]
@Xymanek already done that way :)

File: src/ORM/Exception/RolledbackTransactionException.php [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+use Cake\Core\Exception\Exception;
+
+/**
+ * Used when a behavior cannot be found.
+ *
+ */
+class RolledbackTransactionException extends Exception
+{
+
+    protected $_messageTemplate = 'The afterSave event in "%s" is aborting the transaction before the save process is done.';
```
Comment (ionas) [inline src/ORM/Exception/RolledbackTransactionException.php]: Is this a developer/internal error or something end users should see? If latter: How does it get translated? Can we simply add `__d('cake',` via constructor ..?

File: src/ORM/Exception/RolledbackTransactionException.php [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+ * @copyright     Copyright (c) Cake Software Foundation, Inc. (http://cakefoundation.org)
+ * @since         3.2.13
+ * @license       http://www.opensource.org/licenses/mit-license.php MIT License
+ */
+namespace Cake\ORM\Exception;
+
+use Cake\Core\Exception\Exception;
+
+/**
+ * Used when a behavior cannot be found.
```
Comment (ADmad) [inline src/ORM/Exception/RolledbackTransactionException.php]: This description needs to be updated :)

**Comment by Xymanek** [comment]
@lorenzo I guess I wasn't clear then, my bad. What I'm looking for is the following:

```php
try {
	$table->save();
} catch (DatabaseException $e) {
	$this->Flash->error('Database error');
} finally {
	$this->Flash->success('Something has been saved!');
}
```

So far I've checked the `save()` and it throws a couple of RuntimeExceptions and InvalidArgumentException. The exception that you are adding right now extends base Cake's exception. They don't share a parent so I can't `catch() {}` them all in one statement.

P.S. There is also PDOException that needs to be caught separately, but still...
P.P.S. I just realized that this is outside of the scope of current PR, but I said it here because here is a new exception is added

**Comment by lorenzo** [comment]
@Xymanek This is a developer error, though, not a database error.

**Comment by Xymanek** [comment]
By DatabaseException I mean any exceptions raised in the model layer of the application. If you think that the name is misleading, how about ModelException or something?

File: src/ORM/Exception/RolledbackTransactionException.php (hunk 1 above)
Comment (markstory) [inline src/ORM/Exception/RolledbackTransactionException.php]: If your end users see this, you goofed up imo.

File: src/ORM/Exception/RolledbackTransactionException.php [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+ *
+ */
+class RolledbackTransactionException extends Exception
```
Comment (markstory) [inline src/ORM/Exception/RolledbackTransactionException.php]: To @Xymanek's point we could extend `Cake\Database\Exception` instead?

**Comment by thinkingmedia** [comment]
@Xymanek I agree.

Rename the exception to `TransactionException` so that it can be used again elsewhere, and extend `Cake\Database\Exception`.

**Comment by lorenzo** [comment]
I rebased and fixed the docblock. On the topic of renaming the exception or making it extend another class: This is exception meant for development mode, geared towards preventing a developer error cause because of a limitation in the ORM. It is not an "exceptional" case caused by a database corrupt state (such as a duplicate key  exception).

The intention is that this exception does not get caught by accident and confused with any other possible database error. It need to be solved before the code goes to production.

**Comment by ADmad** [comment]
LGTM

**Comment by markstory** [comment]
Nice work @lorenzo

**Pull request merged**